x-pack/plugins/security_solution/public/timelines/store/middlewares/timeline_note.test.ts

```diff
@@ -0,0 +1,136 @@
+    jest.clearAllMocks();
+  });
+
+  it('should persist a timeline note', async () => {
+    (persistNote as jest.Mock).mockResolvedValue({
+      data: {
+        persistNote: {
+          code: 200,
+          message: 'success',
+          note: {
+            noteId: testNote.id,
+          },
+        },
+      },
+    });
+    expect(selectTimelineById(store.getState(), TimelineId.test).noteIds).toEqual([]);
+    await store.dispatch(updateNote({ note: testNote }));
+    await store.dispatch(addNote({ id: TimelineId.test, noteId: testNote.id }));
+
+    expect(startTimelineSavingMock).toHaveBeenCalled();
+    expect(refreshTimelines as unknown as jest.Mock).toHaveBeenCalled();
+    expect(endTimelineSavingMock).toHaveBeenCalled();
+    expect(selectTimelineById(store.getState(), TimelineId.test).noteIds).toContain(testNote.id);
+  });
+
+  it('should persist a note on an event of a timeline', async () => {
+    (persistNote as jest.Mock).mockResolvedValue({
+      data: {
+        persistNote: {
+          code: 200,
+          message: 'success',
+          note: {
+            noteId: testNote.id,
+          },
+        },
+      },
+    });
+    expect(selectTimelineById(store.getState(), TimelineId.test).eventIdToNoteIds).toEqual({});
+    await store.dispatch(updateNote({ note: testNote }));
+    await store.dispatch(
+      addNoteToEvent({ eventId: testEventId, id: TimelineId.test, noteId: testNote.id })
+    );
+
+    expect(startTimelineSavingMock).toHaveBeenCalled();
+    expect(refreshTimelines as unknown as jest.Mock).toHaveBeenCalled();
+    expect(endTimelineSavingMock).toHaveBeenCalled();
+    expect(selectTimelineById(store.getState(), TimelineId.test).eventIdToNoteIds).toEqual(
+      expect.objectContaining({
+        [testEventId]: [testNote.id],
+      })
+    );
+  });
+
+  it('should show an error message when the call is unauthorized', async () => {
+    (persistNote as jest.Mock).mockResolvedValue({
+      data: {
+        persistNote: {
+          code: 403,
+        },
+      },
+    });
+
+    await store.dispatch(updateNote({ note: testNote }));
+    await store.dispatch(addNote({ id: TimelineId.test, noteId: testNote.id }));
+
+    expect(startTimelineSavingMock).toHaveBeenCalled();
+    expect(endTimelineSavingMock).toHaveBeenCalled();
+    expect(showCallOutUnauthorizedMsgMock).toHaveBeenCalled();
+  });
+
+  it('should show a generic error when the persistence throws', async () => {
+    const addDangerMock = jest.spyOn(kibanaMock.notifications.toasts, 'addDanger');
+    (persistNote as jest.Mock).mockImplementation(() => {
+      throw new Error();
+    });
+
+    await store.dispatch(updateNote({ note: testNote }));
+    await store.dispatch(addNote({ id: TimelineId.test, noteId: testNote.id }));
+
+    expect(startTimelineSavingMock).toHaveBeenCalled();
+    expect(endTimelineSavingMock).toHaveBeenCalled();
+    expect(addDangerMock).toHaveBeenCalled();
+  });
```

Comment:
Not sure if it belongs here or the pinned_events one, but adding a note to an event auto pins the note as well. Do you mind adding a test for that scenario as well?

Reply:
I hadn't heard of that feature before so I was looking for it in the middlewares but it it's actually done by hand:
https://github.com/elastic/kibana/blob/9262d96788dbf23c20f29b8f778accd862d07ca6/x-pack/plugins/security_solution/public/timelines/components/timeline/body/events/stateful_event.tsx#L211-L219

It also appears it's that that entire file isn't tested yet 😨 

I can propose to move this functionality to the reducer where this kind of business logic actually belongs. Should probably do that in a follow-up PR though. Wdyt?

Reply:
Oh, well that's fun and depressing 😅 . Yea, if you can, that'd be great! It'd be nice to avoid introducing functional code changes in this PR